ftp/src/main/scala/akka/stream/alpakka/ftp/impl/FtpBrowserGraphStage.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.ftp
+package impl
+
+import akka.Done
+import akka.stream.stage.{ GraphStageWithMaterializedValue, OutHandler }
+import akka.stream.{ Attributes, Outlet, SourceShape }
+import scala.concurrent.{ Future, Promise }
+
+private[ftp] trait FtpBrowserGraphStage[FtpClient]
+    extends GraphStageWithMaterializedValue[SourceShape[FtpFile], Future[Done]] {
+
+  def name: String
+
+  def basePath: String
+
+  def connectionSettings: RemoteFileSettings
+
+  implicit def ftpClient: FtpClient
+
+  val ftpLike: FtpLike[FtpClient]
+
+  val shape = SourceShape(Outlet[FtpFile](s"$name.out"))
+
+  def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+
+    val matValuePromise = Promise[Done]()
+
+    val logic = new FtpGraphStageLogic[FtpFile, FtpClient](shape, ftpLike, connectionSettings) {
+
+      private var currentFiles: Stream[FtpFile] = Stream.empty[FtpFile]
```

Comment:
If you list a directory with 50 000 files, it would (if I'm right about memoization) keep all of them in memory until the stream completes which isn't very nice.